src/interned.rs

```diff
@@ -0,0 +1,599 @@
+        tables
+            .map
+            .iter()
+            .map(|(key, index)| {
+                TableEntry::new(key.clone(), Some(<Q::Value>::from_u32(index.index)))
+            })
+            .collect()
+    }
+}
+
+impl<DB, Q> QueryStorageMassOps<DB> for InternedStorage<DB, Q>
+where
+    Q: Query<DB>,
+    Q::Value: InternKey,
+    DB: Database,
+{
+    fn sweep(&self, db: &DB, strategy: SweepStrategy) {
+        let mut tables = self.tables.write();
+        let revision_now = db.salsa_runtime().current_revision();
+        let InternTables {
+            map,
+            values,
+            first_free,
+        } = &mut *tables;
+        map.retain(|key, intern_index| {
+            let discard = match strategy.discard_if {
+                DiscardIf::Never => false,
+                DiscardIf::Outdated => match values[intern_index.index()] {
+                    InternValue::Present { accessed_at, .. } => accessed_at < revision_now,
+
+                    InternValue::Free { .. } => {
+                        panic!(
+                            "key {:?} maps to index {:?} which is free",
+                            key, intern_index
+                        );
+                    }
+                },
+                DiscardIf::Always => true,
+            };
+
+            if discard {
+                values[intern_index.index()] = InternValue::Free { next: *first_free };
+                *first_free = Some(*intern_index);
+            }
+
+            !discard
+        });
+    }
+}
+
+impl<DB, Q, IQ> QueryStorageOps<DB, Q> for LookupInternedStorage<DB, Q, IQ>
+where
+    Q: Query<DB>,
+    Q::Key: InternKey,
+    Q::Value: Eq + Hash,
+    IQ: Query<
+        DB,
+        Key = Q::Value,
+        Value = Q::Key,
+        Storage = InternedStorage<DB, IQ>,
+        Group = Q::Group,
+        GroupStorage = Q::GroupStorage,
+        GroupKey = Q::GroupKey,
+    >,
+    DB: Database + HasQueryGroup<Q::Group>,
+{
+    fn try_fetch(
+        &self,
+        db: &DB,
+        key: &Q::Key,
+        database_key: &DB::DatabaseKey,
+    ) -> Result<Q::Value, CycleDetected> {
+        let index = InternIndex::from(key);
+
+        let group_storage = <DB as HasQueryGroup<Q::Group>>::group_storage(db);
+        let interned_storage = IQ::query_storage(group_storage);
+        let StampedValue { value, changed_at } =
+            interned_storage.lookup_value(db, index, Clone::clone);
+
+        db.salsa_runtime()
+            .report_query_read(database_key, changed_at);
+
+        Ok(value)
+    }
+
+    fn maybe_changed_since(
+        &self,
+        db: &DB,
+        revision: Revision,
+        key: &Q::Key,
+        _database_key: &DB::DatabaseKey,
+    ) -> bool {
+        let index = InternIndex::from(key);
+
+        // FIXME -- This seems maybe not quite right, as it will panic
+        // if `key` has been removed from the map since, but it should
+        // return true in that event.
```

Comment:
This should be fixed, right?

As in, gc might trigger a panic.

Reply:
I just documented that. I believe this should be OK.

Reply:
But I might be wrong, let me review the places this is getting called from.

Reply:
Still not entirely convinced that concurrent GC won't cause problems.

Reply:
OK, I did a quick review. I believe a panic is acceptable here. But of course we could try to make it more bullet-proof just because.